client/src/procedures/primitives.rs

```diff
@@ -789,3 +798,115 @@ impl UseSecret for AeadDecrypt {
         &self.key
     }
 }
+
+/// Executes the concat KDF as defined in Section 5.8.1 of NIST.800-56A.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ConcatKdf {
+    /// The hash function to use in the kdf.
+    pub hash: Sha2Hash,
+    /// The identifier of the used algorithm, e.g. `ECDH-ES+A256KW`.
+    pub algorithm_id: String,
+    /// The location of the shared secret `z`.
+    pub shared_secret: Location,
+    /// The number of bytes of key material that should be derived.
+    pub key_len: usize,
+    /// Agreement PartyUInfo
+    pub apu: Vec<u8>,
+    /// Agreement PartyVInfo
+    pub apv: Vec<u8>,
+    /// The location to write the derived key material into.
+    pub output: Location,
+}
+
+impl DeriveSecret for ConcatKdf {
+    type Output = ();
+
+    fn derive(self, guard: Buffer<u8>) -> Result<Products<()>, FatalProcedureError> {
+        let derived_key_material: Vec<u8> = match self.hash {
+            Sha2Hash::Sha256 => Self::concat_kdf(
+                Sha256::new(),
+                self.algorithm_id.as_ref(),
+                self.key_len,
+                guard.borrow().as_ref(),
+                self.apu.as_ref(),
+                self.apv.as_ref(),
+            ),
+            Sha2Hash::Sha384 => Self::concat_kdf(
+                Sha384::new(),
+                self.algorithm_id.as_ref(),
+                self.key_len,
+                guard.borrow().as_ref(),
+                self.apu.as_ref(),
+                self.apv.as_ref(),
+            ),
+            Sha2Hash::Sha512 => Self::concat_kdf(
+                Sha512::new(),
+                self.algorithm_id.as_ref(),
+                self.key_len,
+                guard.borrow().as_ref(),
+                self.apu.as_ref(),
+                self.apv.as_ref(),
+            ),
+        }?;
+
+        Ok(Products {
+            secret: derived_key_material,
+            output: (),
+        })
+    }
+
+    fn source(&self) -> &Location {
+        &self.shared_secret
+    }
+
+    fn target(&self) -> &Location {
+        &self.output
+    }
+}
+
+impl ConcatKdf {
+    /// The Concat KDF as defined in Section 5.8.1 of NIST.800-56A.
+    pub fn concat_kdf<D: Digest>(
```

Comment:
Please add more documentation, a short explanation how the algorithm works and it's parameters.